8131062: aarch64: add support for GHASH acceleration

navyxliu opened the pull request:
Summary: Add support for GHASH using pmull
Reviewed-by: kvn, goetz, aph
Contributed-by: alexander.alexeev@caviumnetworks.com

Thank you for taking the time to help improve OpenJDK and Corretto.

If your pull request concerns a security vulnerability then please do not file it.
Instead, report the problem by email to aws-security@amazon.com.
(You can find more information regarding security issues at https://aws.amazon.com/security/vulnerability-reporting/.)

Otherwise, if your pull request concerns OpenJDK 8
and is not specific to Corretto 8,
then we ask you to redirect your contribution to the OpenJDK project.
See http://openjdk.java.net/contribute/ for details on how to do that.

If your issue is specific to Corretto 8,
then you are in the right place.
Please fill in the following information about your pull request.

### Description
`GHASH.update` is a step in AES/GCM, which authenticates encrypted and additional authenticated data(AAD). The cost of authentication may dominate the AES/GCM cryptographic procedure because:
1.  HotSpot has utilized the dedicated crypto instruction to accelerate AES
2. Applications feed long AAD.  In particular, GMAC is the degeneration form of AES/GCM, which has empty encrypted data.

In one application, we have seen 85% time of AES/GCM kernel(`com.sun.crypto.provider.GaloisCounterMode.doLastBlock`) spends in `GHASH.update` on an ARMv8 instance.  This patch makes use of armv8 NEON instructions to accelerate `GHASH.update`. 

### Related issues
### Motivation and context
We've seen that 4~5 times speedup for the microbenchmarks `AESGCMBench` and `AESGCMByteBuffer` of OpenJDK on AWS EC2 m6g (Graviton2). No regression has been identified by testing.  Here are two representative microbenchmarks. 

Before:
|Benchmark                                               |Mode |Threads|Samples|Score        |Score Error (99.9%)|Unit |Param: dataSize|Param: keyLength|Param: provider|
|--------------------------------------------------------|-----|-------|-------|-------------|-------------------|-----|---------------|----------------|---------------|
|o.o.b.j.c.s.AESGCMBench.decrypt                         |thrpt|1      |40     |49386.026271 |88.056237          |ops/s|1024           |128             |               |
|o.o.b.j.c.s.AESGCMBench.encrypt                         |thrpt|1      |40     |50810.762489 |66.829595          |ops/s|1024           |128             |               |

After
|Benchmark                                               |Mode |Threads|Samples|Score        |Score Error (99.9%)|Unit |Param: dataSize|Param: keyLength|Param: provider|
|--------------------------------------------------------|-----|-------|-------|-------------|-------------------|-----|---------------|----------------|---------------|
|o.o.b.j.c.s.AESGCMBench.decrypt|thrpt|1      |40     |237779.050064|1466.294778        |ops/s|1024           |128             |               |
|o.o.b.j.c.s.AESGCMBench.encrypt|thrpt|1      |40     |256845.947199|1599.024270        |ops/s|1024           |128             |               |


### How has this been tested?
I have jtreg test subset: `jdk_security2` of openjdk8u-dev, which covers javax.crypto.* 

`
Summary: jdk_security2
TEST STATS: name=jdk_security2 run=177 pass=177 fail=0
`

The following two tests are directly related among them.
```
com/sun/crypto/provider/Cipher/AEAD/GCMLargeDataKAT.java                             Passed. Execution successful
com/sun/crypto/provider/Cipher/AES/TestKATForGCM.java                                Passed. Execution successful
```

### Platform information
    Works on armv8


### Additional context



### Contribution confirmation

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Commit:
8131062: aarch64: add support for GHASH acceleration

Summary: Add support for GHASH using pmull
Reviewed-by: kvn, goetz, aph
Contributed-by: alexander.alexeev@caviumnetworks.com

phohensee commented on src/hotspot/src/cpu/aarch64/vm/assembler_aarch64.hpp:
```diff
@@ -1,6 +1,6 @@
 /*
- * Copyright (c) 1997, 2019, Oracle and/or its affiliates. All rights reserved.
- * Copyright (c) 2013, 2019, Red Hat Inc. All rights reserved.
+ * Copyright (c) 1997, 2021, Oracle and/or its affiliates.
```
No need for a copyright update for backports. This one is dated 2015, so the existing 2019 is fine.

navyxliu replied:
I update copyright headers.

phohensee commented on src/hotspot/src/cpu/aarch64/vm/stubGenerator_aarch64.cpp:
```diff
@@ -1,6 +1,6 @@
 /*
+ * Copyright (c) 2003, 2021, Oracle and/or its affiliates.
```
Copyright date should be updated to 2015, not 2021.

phohensee commented on src/hotspot/src/cpu/aarch64/vm/vm_version_aarch64.cpp:
```diff
@@ -1,6 +1,6 @@
 /*
- * Copyright (c) 2013, Red Hat Inc.
- * Copyright (c) 1997, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1997, 2021, Oracle and/or its affiliates.
```
No need for a copyright date update.

phohensee reviewed (changes requested):
Looks good except for copyright date adjustments.

eastig reviewed (commented):
Besides Paul's comments, the rest looks good to me.

Commit:
Fix copyright years.

The original patch happened in 2015, so only update copyright years
to 2015 in the backport.
